File: BudgetPro.Server/Services/BudgetService.cs
using BudgetPro.Server.Data;
using BudgetPro.Server.DTOs;
using BudgetPro.Server.Interfaces;
using BudgetPro.Server.Requests.AddRequests;
using BudgetPro.Server.Requests.GetRequests;
using Microsoft.EntityFrameworkCore;

namespace BudgetPro.Server.Services;

public class BudgetService : IBudgetService
{
    private readonly IDbContextFactory<BudgetDbContext> dbContextFactory;
    public BudgetService(IDbContextFactory<BudgetDbContext> dbContextFactory)
    {
        this.dbContextFactory = dbContextFactory;
    }

    public async Task<BudgetDTO> AddNewBudget(AddBudgetRequest request)
    {
        if (request is null)
        {
            return new BudgetDTO();
        }

        using var context = await dbContextFactory.CreateDbContextAsync();

        var possibleBudget = await context.Budgets
            .Where(x => x.Name == request.Name && x.CategoryId == request.CategoryId)
            .FirstOrDefaultAsync();

        if (possibleBudget is not null)
        {
            return new BudgetDTO();
        }

        Budget newBudget = new Budget()
        {
            Amount = request.Amount,
            CategoryId = request.CategoryId,
            Name = request.Name,
            Timeframe = request.TimeFrame,
            UserId = request.UserId
        };

        await context.AddAsync(newBudget);
        await context.SaveChangesAsync();

        return newBudget.ToDTO();
    }

    public async Task<List<BudgetDTO>> GetAllBudgets()
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        return await context.Budgets
            .Include(x => x.User)
            .Include(x => x.Category)
            .Select(x => x.ToDTO())
            .ToListAsync();
    }

    public async Task<BudgetDTO> GetBudgetById(FindBudgetRequest request)
    {
        if (request is null || request.Id <= 0)
        {
            return new BudgetDTO();
        }

        using var context = await dbContextFactory.CreateDbContextAsync();

        return await context.Budgets
            .Include(x => x.User)
            .Include(x => x.Category)
            .Where(x => x.Id == request.Id)
            .Select(x => x.ToDTO())
            .FirstOrDefaultAsync() ?? new BudgetDTO();
    }

    public async Task<BudgetDTO> GetBudgetByName(FindBudgetRequest request)
    {
        if (request is null || request.Name is null)
        {
            return new BudgetDTO();
        }

        using var context = await dbContextFactory.CreateDbContextAsync();

        return await context.Budgets
            .Include(x => x.User)
            .Include(x => x.Category)
            .Where(x => x.Name == request.Name)
            .Select(x => x.ToDTO())
            .FirstOrDefaultAsync() ?? new BudgetDTO();
    }

    public async Task<BudgetDTO> GetBudgetByTimeFrame(FindBudgetRequest request)
    {
        if (request is null || request.TimeFrame is null || request.UserId <= 0)
        {
            return new BudgetDTO();
        }

        using var context = await dbContextFactory.CreateDbContextAsync();

        return await context.Budgets
            .Include(x => x.User)
            .Include(x => x.Category)
            .Where(x => x.UserId == request.UserId && x.Timeframe == request.TimeFrame)
            .Select(x => x.ToDTO())
            .FirstOrDefaultAsync() ?? new BudgetDTO();
    }

    public async Task<List<BudgetDTO>> GetBudgetsByCategory(FindBudgetRequest request)
    {
        if (request is null || request.CategoryId <= 0)
        {
            return new List<BudgetDTO>();
        }

        using var context = await dbContextFactory.CreateDbContextAsync();

        return await context.Budgets
            .Include(x => x.User)
            .Include(x => x.Category)
            .Where(x => x.CategoryId == request.CategoryId)
            .Select(x => x.ToDTO())
            .ToListAsync();
    }

    public async Task<List<BudgetDTO>> GetBudgetsByUser(FindBudgetRequest request)
    {
        if (request is null || request.UserId <= 0)
        {
            return new List<BudgetDTO>();
        }

        using var context = await dbContextFactory.CreateDbContextAsync();

        return await context.Budgets
            .Include(x => x.User)
            .Include(x => x.Category)
            .Where(x => x.UserId == request.UserId)
            .Select(x => x.ToDTO())
            .ToListAsync();
    }
}

File: budgetpro.client/src/app/services/budget/budget.service.ts
import { Injectable } from '@angular/core';
import { Category } from '../category/category.service';
import { User } from '../auth/auth.service';
import { environment } from '../../../environments/environment';
import { HttpClient } from '@angular/common/http';

export interface Budget {
  id: number;
  userId: number;
  categoryId: number;
  amount: number;
  endDate: Date;
  startDate: Date;
  name: string;
  category: Category;
  user: User;
}

export interface AddBudgetRequest {
  userId: number;
  categoryId: number;
  amount: number;
  endDate: Date;
  startDate: Date;
  name: string;
}

export interface FindBudgetRequest {
  id: number;
  userId: number;
  categoryId: number;
  endDate: Date;
  startDate: Date;
  name: string;
}

@Injectable({
  providedIn: 'root',
})
export class BudgetService {
  private apiUrl = environment.apiUrl + '/Budget';

  constructor(private httpClient: HttpClient) {}

  getAllBudgets() {
    return this.httpClient.get<Budget[]>(`${this.apiUrl}/GetAll`);
  }

  addNewBudget(request: AddBudgetRequest) {
    return this.httpClient.post<Budget>(`${this.apiUrl}/Add`, request);
  }

  getBudgetById(request: FindBudgetRequest) {
    return this.httpClient.post<Budget>(`${this.apiUrl}/GetById`, request);
  }

  getBudgetsByUser(request: FindBudgetRequest) {
    return this.httpClient.post<Budget[]>(`${this.apiUrl}/GetByUser`, request);
  }

  getBudgetsByCategory(request: FindBudgetRequest) {
    return this.httpClient.post<Budget[]>(
      `${this.apiUrl}/GetByCategory`,
      request
    );
  }

  getBudgetByTimeFrame(request: FindBudgetRequest) {
    return this.httpClient.post<Budget>(
      `${this.apiUrl}/GetByTimeFrame`,
      request
    );
  }

  getBudgetByName(request: FindBudgetRequest) {
    return this.httpClient.post<Budget>(`${this.apiUrl}/GetByName`, request);
  }
}
